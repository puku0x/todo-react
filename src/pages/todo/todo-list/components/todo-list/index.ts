export * from './todo-list.component';
