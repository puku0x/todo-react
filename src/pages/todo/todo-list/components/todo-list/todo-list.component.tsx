import { cx } from '@emotion/css';
import styled from '@emotion/styled';
import * as React from 'react';
import { ChangeEvent, useCallback } from 'react';
import { NavLink as NavLinkBase } from 'react-router-dom';

import { Todo } from '../../../../../models';

const NavLink = styled(NavLinkBase)`
  &.completed {
    text-decoration: line-through;
  }
`;

type Props = {
  isFetching?: boolean;
  todos: Todo[];
  offset: number;
  limit: number;
  onChangeOffset?: (offset: number) => void;
  onChangeLimit?: (limit: number) => void;
};

export const TodoListComponent: React.FC<Props> = React.memo((props) => {
  const { todos, offset, limit, onChangeOffset, onChangeLimit } = props;

  const changeOffset = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      const value = Number(event.target.value);
      onChangeOffset?.(value);
    },
    [onChangeOffset]
  );

  const changeLimit = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      const value = Number(event.target.value);
      onChangeLimit?.(value);
    },
    [onChangeLimit]
  );

  return (
    <>
      <h2>todo-list</h2>
      <p>
        <NavLink to="/todos/new">Add a new todo</NavLink>
      </p>
      <form>
        {' offset '}
        <input
          type="number"
          min="0"
          name="offset"
          defaultValue={offset}
          onChange={changeOffset}
        />
        {' limit '}
        <input
          type="number"
          min="0"
          name="limit"
          defaultValue={limit}
          onChange={changeLimit}
        />
      </form>
      <ul>
        {todos.map((todo) => (
          <li key={todo.id}>
            <NavLink
              className={cx([{ completed: todo.completed }])}
              to={`/todos/${todo.id}`}
            >
              {todo.title}
            </NavLink>
          </li>
        ))}
      </ul>
    </>
  );
});
