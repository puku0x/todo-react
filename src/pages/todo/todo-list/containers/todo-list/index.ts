export * from './todo-list.container';
