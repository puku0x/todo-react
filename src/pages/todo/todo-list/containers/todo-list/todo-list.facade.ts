import { AnyAction, ThunkDispatch, unwrapResult } from '@reduxjs/toolkit';
import { useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useLocation } from 'react-router-dom';

import {
  fetchAllTodos,
  isFetchingSelector,
  todosSelector,
} from '../../../store';

export const useTodoListFacade = (arg: { offset?: number; limit?: number }) => {
  const { offset, limit } = arg;
  const history = useHistory();
  const location = useLocation();
  const dispatch = useDispatch<ThunkDispatch<unknown, undefined, AnyAction>>();
  const isFetching = useSelector(isFetchingSelector);
  const todos = useSelector(todosSelector);

  const fetchAll = useCallback(
    (arg: { offset?: number; limit?: number }) => {
      return dispatch(fetchAllTodos(arg)).then(unwrapResult);
    },
    [dispatch]
  );

  const changeOffset = useCallback(
    (offset: number) => {
      const params = new URLSearchParams(location.search);
      params.set('offset', `${offset}`);
      history.push(`/todos?${params}`);
    },
    [history, location.search]
  );

  const changeLimit = useCallback(
    (limit: number) => {
      const params = new URLSearchParams(location.search);
      params.set('limit', `${limit}`);
      history.push(`/todos?${params}`);
    },
    [history, location.search]
  );

  useEffect(() => {
    fetchAll({ offset, limit });
  }, [offset, limit, fetchAll]);

  return {
    isFetching,
    todos,
    changeOffset,
    changeLimit,
    fetchAll,
  } as const;
};
