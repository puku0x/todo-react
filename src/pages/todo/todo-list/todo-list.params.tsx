import { useLocation } from 'react-router-dom';

export const useTodoListParams = () => {
  const location = useLocation();
  const params = new URLSearchParams(location.search);
  const limitParam = params.get('limit') || '10';
  const offsetParam = params.get('offset') || '0';

  return {
    limit: +limitParam,
    offset: +offsetParam,
  } as const;
};
