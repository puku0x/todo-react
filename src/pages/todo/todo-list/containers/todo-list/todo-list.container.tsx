import { memo } from 'react';

import { TodoList } from '../../components';
import { useTodoListFacade } from './todo-list.facade';

type Props = {
  offset: number;
  limit: number;
};

export const TodoListContainer = memo((props: Props) => {
  const { offset, limit } = props;
  const { isFetching, todos, changeOffset, changeLimit } = useTodoListFacade({
    offset,
    limit,
  });

  return (
    <TodoList
      isFetching={isFetching}
      todos={todos}
      offset={offset}
      limit={limit}
      onChangeOffset={changeOffset}
      onChangeLimit={changeLimit}
    />
  );
});
