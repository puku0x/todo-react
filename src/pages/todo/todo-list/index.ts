import { lazy } from 'react';

export const TodoListPage = lazy(() =>
  import('./todo-list.page').then((m) => ({ default: m.TodoListPage }))
);
