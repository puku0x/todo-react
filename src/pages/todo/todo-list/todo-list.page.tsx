import { FunctionComponent, memo, useMemo } from 'react';
import { useLocation } from 'react-router-dom';

import { TodoListContainer } from './containers';

export const TodoListPage: FunctionComponent = memo(() => {
  const location = useLocation();
  const params = useMemo(() => {
    return new URLSearchParams(location.search);
  }, [location.search]);
  const offset = +(params.get('offset') || '0');
  const limit = +(params.get('limit') || '10');

  return <TodoListContainer offset={offset} limit={limit} />;
});
