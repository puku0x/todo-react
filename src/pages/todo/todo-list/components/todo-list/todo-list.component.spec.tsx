import { act, render } from '@testing-library/react';
import { Fragment } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { createTodosMock } from '../../../../../models/testing';
import { TodoList } from './todo-list.component';

describe('TodoList', () => {
  it('render', async () => {
    const todos = createTodosMock();
    const offset = 0;
    const limit = 10;
    let result = render(<Fragment />);
    await act(async () => {
      result = render(
        <TodoList todos={todos} offset={offset} limit={limit} />,
        { wrapper: MemoryRouter }
      );
    });
    const { asFragment } = result;

    expect(asFragment()).toMatchSnapshot();
  });

  it('render with isFetching', async () => {
    const offset = 0;
    const limit = 10;
    let result = render(<Fragment />);
    await act(async () => {
      result = render(
        <TodoList isFetching todos={[]} offset={offset} limit={limit} />,
        { wrapper: MemoryRouter }
      );
    });
    const { asFragment } = result;

    expect(asFragment()).toMatchSnapshot();
  });
});
