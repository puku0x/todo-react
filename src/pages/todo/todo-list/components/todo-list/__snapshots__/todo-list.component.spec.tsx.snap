// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList render 1`] = `
<DocumentFragment>
  <h2>
    todo-list
  </h2>
  <p>
    <a
      class="css-1qjclwt"
      href="/todos/new"
    >
      Add a new todo
    </a>
  </p>
  <form>
     offset 
    <input
      min="0"
      name="offset"
      type="number"
      value="0"
    />
     limit 
    <input
      min="0"
      name="limit"
      type="number"
      value="10"
    />
  </form>
  <ul>
    <li>
      <a
        class=" css-1qjclwt"
        href="/todos/1"
      >
        title
      </a>
    </li>
    <li>
      <a
        class=" css-1qjclwt"
        href="/todos/2"
      >
        title
      </a>
    </li>
    <li>
      <a
        class=" css-1qjclwt"
        href="/todos/3"
      >
        title
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`TodoList render with isFetching 1`] = `
<DocumentFragment>
  <h2>
    todo-list
  </h2>
  <p>
    <a
      class="css-1qjclwt"
      href="/todos/new"
    >
      Add a new todo
    </a>
  </p>
  <form>
     offset 
    <input
      min="0"
      name="offset"
      type="number"
      value="0"
    />
     limit 
    <input
      min="0"
      name="limit"
      type="number"
      value="10"
    />
  </form>
  <ul />
</DocumentFragment>
`;
